libcoap/src/types.rs

```diff
@@ -258,153 +270,567 @@ impl From<coap_uri_scheme_t> for CoapUriScheme {
-        self.port
+    pub fn port(&self) -> Option<UriPort> {
+        match self.raw_uri.port {
+            0 => None,
+            v => Some(v),
+        }
     }
 
-    /// Drains the parts of the URI path of this CoapUri into an iterator.
-    pub(crate) fn drain_path_iter(&mut self) -> Option<Drain<String>> {
-        self.path.as_mut().map(|p| p.drain(..))
+    /// Returns the URI path part of this URI.
+    pub fn path(&self) -> Option<&[u8]> {
+        let raw_str = self.raw_uri.path;
+        if raw_str.s.is_null() {
+            return None;
+        }
+        // SAFETY: After construction the fields of self.raw_uri always reference the corresponding
+        //         parts of the underlying string, which is pinned. Therefore, the pointer and
+        //         length are valid for the lifetime of this struct.
+        Some(unsafe { std::slice::from_raw_parts(raw_str.s, raw_str.length) })
     }
 
-    /// Returns an iterator over the path components of this URI.
-    pub fn path_iter(&self) -> Option<Iter<'_, String>> {
-        self.path.as_ref().map(|p| p.iter())
+    /// Returns the host part of this URI.
+    pub fn query(&self) -> Option<&[u8]> {
+        let raw_str = self.raw_uri.query;
+        if raw_str.s.is_null() {
+            return None;
+        }
+        // SAFETY: After construction the fields of self.raw_uri always reference the corresponding
+        //         parts of the underlying string, which is pinned. Therefore, the pointer and
+        //         length are valid for the lifetime of this struct.
+        Some(unsafe { std::slice::from_raw_parts(raw_str.s, raw_str.length) })
     }
 
-    /// Drains the parts of the URI query of this CoapUri into an iterator.
-    pub fn drain_query_iter(&mut self) -> Option<Drain<String>> {
-        self.query.as_mut().map(|p| p.drain(..))
+    /// Returns whether this URI is a proxy URI.
+    pub fn is_proxy(&self) -> bool {
+        self.is_proxy
     }
 
-    /// Returns an iterator over the query components of this URI.
-    pub fn query_iter(&self) -> Option<Iter<String>> {
-        self.query.as_ref().map(|p| p.iter())
+    /// Converts the given URI into a `Vec` of [CoapOption]s that can be added to a
+    /// [crate::message::CoapMessage].
+    pub fn into_options(self) -> Vec<CoapOption> {
+        // TODO this is a lot of copying around, however, fixing that would require an entire
+        //      rewrite of the option handling code, so it's better kept for a separate PR.
+        let mut buf = vec![0u8; self.path().map(|v| v.len()).unwrap_or(0) + self.query().map(|v| v.len()).unwrap_or(0)];
+        let mut optlist = std::ptr::null_mut();
+        // SAFETY: self.raw_uri is always valid after construction. The destination may be a null
+        //         pointer, optlist may be a null pointer at the start (it will be set to a valid
+        //         pointer by this call. Buf and create_port_host_opt are set according to the
```

Comment:
```suggestion
        //         pointer, optlist may be a null pointer at the start (it will be set to a valid
        //         pointer by this call). Buf and create_port_host_opt are set according to the
```